File: merge-k-sorted-lists/merge-k-sorted-lists.py
# Definition for single-linked list.
class ListNode(object):
    def __init__(self,x):
        self.val = x
        self.next = None
class Solution(object):
    def mergeKLists(self,lists):
        """

        :param lists: List[ListNode]
        :return: ListNode
        """
        if len(lists)==0:
            return None
        if len(lists)==1:
            return lists[0]
        l1 = self.mergeKLists(lists[:len(lists)/2])
        l2 = self.mergeKLists(lists[len(lists)/2:])
        head = self.merge2Lists(l1,l2)
        return head

    def merge2Lists(self,la,lb):
        """

        :param la: ListNode
        :param lb: ListNode
        :return: ListNode
        """
        if la is None:
            return lb
        elif lb is None:
            return la
        else:
            p = ListNode(0)
            dummyhead = p
            while la is not None and lb is not None:
                if la.val < lb.val:
                    p.next = la
                    la = la.next
                else:
                    p.next = lb
                    lb = lb.next
                p = p.next

            if la is not None:
                p.next = la
            else:
                p.next = lb
        return dummyhead.next

File: 023.merge-k-sorted-lists/merge-k-sorted-lists.cpp
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        ListNode* res = nullptr;
        for(auto it:lists){
            res = merge(res,it);
        }
        return res;
    }
    ListNode* merge2Lists(ListNode* a,ListNode* b){

        ///@note nullptr pointer
        ListNode* res = nullptr;
        if(a== nullptr){
            return b;
        } else if(b== nullptr){
            return a;
        } else {
            if(a->val<b->val){
                res = a;
                a = a->next;
            } else{
                res = b;
                b = b->next;
            }
            ListNode* tmp = res;
            while (a!= nullptr&&b!= nullptr){
                if(a->val<b->val){
                    tmp->next = a;
                    a = a->next;
                } else{
                    tmp->next = b;
                    b = b->next;
                }
                tmp = tmp->next;
            }
            if(a!= nullptr){
                tmp->next = a;
            }else{
                tmp->next = b;
            }
            return res;
        }

    }
    ListNode* merge(ListNode* l1, ListNode* l2){
        if(!l1) return l2;
        if(!l2) return l1;

        if(l1->val < l2->val){
            l1->next = merge(l1->next, l2);
            return l1;
        }else{
            l2->next = merge(l1, l2->next);
            return l2;
        }
    }
};